samza-azure/src/main/java/org/apache/samza/system/azureblob/avro/AzureBlobOutputStream.java

```diff
@@ -0,0 +1,314 @@
+  private final int maxBlockFlushThresholdSize;
+  private final AzureBlobWriterMetrics metrics;
+  private final Compression compression;
+
+  private volatile boolean isClosed = false;
+  private long totalUploadedBlockSize = 0;
+  private int blockNum;
+
+  public AzureBlobOutputStream(BlockBlobAsyncClient blobAsyncClient, Executor blobThreadPool, AzureBlobWriterMetrics metrics,
+      long flushTimeoutMs, int maxBlockFlushThresholdSize, Compression compression) {
+    byteArrayOutputStream = Optional.of(new ByteArrayOutputStream(maxBlockFlushThresholdSize));
+    this.blobAsyncClient = blobAsyncClient;
+    blockList = new ArrayList<>();
+    blockNum = 0;
+    this.blobThreadPool = blobThreadPool;
+    this.flushTimeoutMs = flushTimeoutMs;
+    this.maxBlockFlushThresholdSize = maxBlockFlushThresholdSize;
+    this.metrics = metrics;
+    this.compression = compression;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public synchronized void write(int b) {
+    if (!byteArrayOutputStream.isPresent()) {
+      throw new IllegalStateException("Internal Buffer must have been released earlier for blob " + blobAsyncClient.getBlobUrl().toString());
+    }
+
+    if (byteArrayOutputStream.get().size() + 1 > maxBlockFlushThresholdSize) {
+      uploadBlockAsync();
+    }
+    byteArrayOutputStream.get().write(b);
+    metrics.updateWriteByteMetrics(1);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public synchronized void write(byte[] b, int off, int len) {
+    if (!byteArrayOutputStream.isPresent()) {
+      throw new IllegalStateException("Internal Buffer must have been released earlier for blob " + blobAsyncClient.getBlobUrl().toString());
+    }
+
+    int remainingBytes = len;
+    int offset = off;
+    while (remainingBytes > 0) {
+      int bytesToWrite = Math.min(maxBlockFlushThresholdSize - byteArrayOutputStream.get().size(), remainingBytes);
+      byteArrayOutputStream.get().write(b, offset, bytesToWrite);
+      offset += bytesToWrite;
+      remainingBytes -= bytesToWrite;
+      if (byteArrayOutputStream.get().size() >= maxBlockFlushThresholdSize) {
+        uploadBlockAsync();
+      }
+    }
+    metrics.updateWriteByteMetrics(len);
+  }
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public synchronized void flush() {
+    if (byteArrayOutputStream.isPresent()) {
+      uploadBlockAsync();
+    }
+  }
+
+  /**
+   * This api waits for all pending upload (stageBlock task) futures to finish.
+   * It then synchronously commits the list of blocks to persist the actual blob on storage.
+   * @throws IllegalStateException when
+   *       - when closing an already closed stream
+   * @throws RuntimeException when
+   *       - byteArrayOutputStream.close fails or
+   *       - any of the pending uploads fails or
+   *       - blob's commitBlockList fails
+   */
+  @Override
+  public synchronized void close() {
+
+    if (isClosed) {
+      return;
```

Comment:
This method is going to be called twice in the event of no failures - once by DataFileWriter.close and in the finally block of AzureBlobAvroWriter.close - This was done to avoid loss of blocks if DataFilwWriter.close failed. Hence, logging something would unnecessarily confuse the users

Reply:
Hm, generally it is not the case that a `close` is expected to be called twice (and then ignored the second time). Is it difficult to only have it called once in the happy path case?